### PR TITLE
adds local sanity tests for lists

### PR DESCRIPTION
**Pull request by HiteshRepo**
adds local sanity tests for lists

for reference:
- `lists` do not have `folders` concept, they are directly created in a site. As a result cannot be restored to a different folder to test. Therefore they need to be created in adjacent to the 'backed-up' lists. Due to this there is a requirement of `SANITY_TEST_RESTORE_CONTAINER_PREFIX`. Using this, during restore check, this env-var is used to find restored lists.
- Similarly, `SANITY_TEST_SOURCE_CONTAINER` is also used a prefix matcher to filter created test lists
- When exported, lists are stored in local into the provided directory as `<SANITY_TEST_RESTORE_CONTAINER>/Lists/<list-id>/<list-id>.json`

should go in after:
- https://github.com/alcionai/corso/pull/5024
- https://github.com/alcionai/corso/pull/5026

#### Does this PR need a docs update or release note?
- [x] :no_entry: No

#### Type of change

- [x] :robot: Supportability/Tests

#### Issue(s)
#4754

#### Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/cmd/sanity_test/common/filepath.go [hunk 1]
```diff
@@ -78,3 +68,104 @@ func BuildFilepathSanitree(
 
 	return root
 }
+
+func BuildFilepathSanitreeForSharepointLists(
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/common/filepath.go:72]: this should live in /export/sharepoint.  It is not a common (ie: generic) func.

File: src/cmd/sanity_test/export/sharepoint.go [hunk 2]
```diff
@@ -14,15 +20,65 @@ func CheckSharePointExport(
 	ac api.Client,
 	envs common.Envs,
 ) {
-	drive, err := ac.Sites().GetDefaultDrive(ctx, envs.SiteID)
-	if err != nil {
-		common.Fatal(ctx, "getting the drive:", err)
+	if envs.Category == "lists" {
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/export/sharepoint.go:23]: category checks should compare against `path` package consts (or their string values) for consistency.  Avoid hardcoded strings if possible.

File: src/cmd/sanity_test/export/sharepoint.go [hunk 3]
```diff
@@ -14,15 +20,65 @@ func CheckSharePointExport(
...
+			envs)
+	}
+}
+
+func CheckSharepointListsExport(
+	ctx context.Context,
+	ac api.Client,
+	envs common.Envs,
+) {
+	exportFolderName := "Lists"
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/export/sharepoint.go:47]: The `.Human()` value of the lists category would work better here than a hardcoded string.

File: src/cmd/sanity_test/common/common.go [hunk 4]
```diff
@@ -19,33 +19,40 @@ type PermissionInfo struct {
...
-	sanityBackupID             = "SANITY_BACKUP_ID"
-	sanityTestSourceContainer  = "SANITY_TEST_SOURCE_CONTAINER"
-	sanityTestRestoreContainer = "SANITY_TEST_RESTORE_CONTAINER"
-	sanityTestUser             = "SANITY_TEST_USER"
+	sanityBackupID                   = "SANITY_BACKUP_ID"
+	sanityTestSourceContainer        = "SANITY_TEST_SOURCE_CONTAINER"
+	sanityTestRestoreContainer       = "SANITY_TEST_RESTORE_CONTAINER"
+	sanityTestRestoreContainerPrefix = "SANITY_TEST_RESTORE_CONTAINER_PREFIX"
+	sanityTestUser                   = "SANITY_TEST_USER"
+	sanityTestCategory               = "SANITY_TEST_CAT"
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/common/common.go:27]: Nit: no need to shorten the word.
```suggestion
	sanityTestCategory               = "SANITY_TEST_CATEGORY"
```

File: src/cmd/sanity_test/restore/sharepoint.go [hunk 5]
```diff
@@ -13,16 +19,150 @@ func CheckSharePointRestoration(
...
+	site, err := ac.Sites().GetByID(ctx, siteID, api.CallConfig{})
+	if err != nil {
+		common.Fatal(
+			ctx,
+			fmt.Sprintf("finding site by id %q", siteID),
+			err)
+	}
+
+	cfg := api.CallConfig{
+		Select: []string{"id", "displayName", "list", "lastModifiedDateTime"},
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/restore/sharepoint.go:83]: should use `IDAnd(...)` here, for consistency.

File: src/cmd/sanity_test/restore/sharepoint.go [hunk 6]
```diff
@@ -13,16 +19,150 @@ func CheckSharePointRestoration(
...
+			Size: int64(len(listItems)),
+		}
+
+		root.Leaves[m.ID] = m
+	}
+
+	return root
+}
+
+func getAllowedLists(lists []models.Listable) []models.Listable {
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/restore/sharepoint.go:139]: nit: `get` makes it sound like you're retrieving a value from a database or external server.  Prefer `filter` for funcs like this.
```suggestion
func filterToSupportedLists(lists []models.Listable) []models.Listable {
```

File: src/cmd/sanity_test/restore/sharepoint.go [hunk 7]
```diff
@@ -13,16 +19,150 @@ func CheckSharePointRestoration(
...
+
+	for _, list := range lists {
+		listDisplayName := ptr.Val(list.GetDisplayName())
+
+		if strings.HasPrefix(listDisplayName, prefix) {
+			filteredLists = append(filteredLists, list)
+		}
+	}
+
+	return filteredLists
```
Comment (ryanfkeepers) [inline src/cmd/sanity_test/restore/sharepoint.go:167]: nit: this func is overcomplicating things by acting recursively when it doesn't need recursion.
```suggestion
func filterListsByPrefix(lists, prefix string) []models.Listable {
        result := []models.Listable{}

	for _, list := range lists {
	        for _, pfx := range strings.Split(prefix, ",") {
		    if strings.HasPrefix(ptr.Val(list.GetDisplayName()), pfx) {
			    result = append(result, list)
			    break
		    }
		}
	}

	return result
```

**Review by ryanfkeepers — APPROVED** [review]
generally looks good, just some nits.  Have you verified the success of this test manually yet?

**Comment by HiteshRepo** [comment]
> generally looks good, just some nits. Have you verified the success of this test manually yet?

yes @ryanfkeepers and the CI sanity tests are built on top of this PR.

**Pull request merged**